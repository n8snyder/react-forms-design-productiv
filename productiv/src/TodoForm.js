import React, { useState } from "react";

/** Form for adding.
 *
 * Props:
 * - initialFormData
 * - handleSave: function to call in parent.
 * 
 * State:
 * - formData: { title: '', description: '', priority: 1 }
 *
 * { TodoApp, EditableTodo } -> TodoForm
 */

function TodoForm({ handleSave, initialFormData =
  { title: '', description: '', priority: 1 } }) {
  const [formData, setForm] = useState(initialFormData);

  /** Update form input. */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setForm(oldForm => {
      return { ...oldForm, [name]: value };
    });
  }

  /** Call parent function and clear form. */
  function handleSubmit(evt) {
    evt.preventDefault();
    handleSave(formData);
    setForm(initialFormData);
  }

  return (
    <form className="TodoForm" onSubmit={handleSubmit}>

      <div className="mb-3">
        <input
          id="TodoForm-title"
          name="title"
          className="form-control"
          placeholder="Title"
          onChange={handleChange}
          value={formData.title}
          aria-label="Title"
        />
      </div>

      <div className="mb-3">
        <textarea
          id="TodoForm-description"
          name="description"
          className="form-control"
          placeholder="Description"
          onChange={handleChange}
          value={formData.description}
          aria-label="Description"
        />
      </div>

      <div className="mb-3 d-flex justify-content-between">
        <div className="w-75 d-flex justify-content-between">
          <label htmlFor="newTodo-priority"
            className="d-inline-flex">Priority:&nbsp;&nbsp;
          </label>
          <select id="TodoForm-priority"
            name="priority"
            value={formData.priority}
            onChange={handleChange}
            className="form-control form-control-sm d-inline-flex"
          >
            <option value={1}>Ultra-Über</option>
            <option value={2}>Über</option>
            <option value={3}>Meh</option>
          </select>
        </div>
        <button className="btn-primary rig btn btn-sm NewTodoForm-addBtn">
          Gø!
        </button>
      </div>

    </form>
  );
}

export default TodoForm;
