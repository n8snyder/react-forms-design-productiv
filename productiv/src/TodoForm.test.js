import TodoForm from './TodoForm';
import { render, fireEvent } from "@testing-library/react";
const mockHandleSave = jest.fn();

const initalState = { title: 'testTitle', description: 'testDescription', priority: 1 };

describe("Todo", function () {
  it("renders without crashing", function () {
    render(<TodoForm
      handleSave={mockHandleSave}
      initialFormData={initalState} />);
  });

  it("snapshot", function () {
    const { container } = render(<TodoForm
      handleSave={mockHandleSave}
      initialFormData={initalState} />);
    expect(container).toMatchSnapshot();
  });

  it("submitting the form", function () {
    const { container } = render(<TodoForm
      handleSave={mockHandleSave}
      initialFormData={initalState} />);
    expect(mockHandleSave).toHaveBeenCalledTimes(0);

    //clicking the button to submit form
    const editForm = container.querySelector(".TodoForm");
    fireEvent.submit(editForm);
    expect(mockHandleSave).toHaveBeenCalledTimes(1);
  });

  it("initial form data", function () {
    const { container, debug } = render(<TodoForm
      handleSave={mockHandleSave}
      initialFormData={initalState} />);

    // debug(container.querySelector("#TodoForm-priority"));

    expect(container.querySelector("#TodoForm-title")
      .getAttribute("value"))
      .toEqual("testTitle");

    expect(container.querySelector("#TodoForm-description"))
      .toContainHTML("testDescription");

    expect(container.querySelector("#TodoForm-priority").selectedIndex)
      .toEqual(0);
  });
});