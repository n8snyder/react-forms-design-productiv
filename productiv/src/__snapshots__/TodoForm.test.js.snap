// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo snapshot 1`] = `
<div>
  <form
    class="TodoForm"
  >
    <div
      class="mb-3"
    >
      <input
        aria-label="Title"
        class="form-control"
        id="TodoForm-title"
        name="title"
        placeholder="Title"
        value="testTitle"
      />
    </div>
    <div
      class="mb-3"
    >
      <textarea
        aria-label="Description"
        class="form-control"
        id="TodoForm-description"
        name="description"
        placeholder="Description"
      >
        testDescription
      </textarea>
    </div>
    <div
      class="mb-3 d-flex justify-content-between"
    >
      <div
        class="w-75 d-flex justify-content-between"
      >
        <label
          class="d-inline-flex"
          for="newTodo-priority"
        >
          Priority:  
        </label>
        <select
          class="form-control form-control-sm d-inline-flex"
          id="TodoForm-priority"
          name="priority"
        >
          <option
            value="1"
          >
            Ultra-Über
          </option>
          <option
            value="2"
          >
            Über
          </option>
          <option
            value="3"
          >
            Meh
          </option>
        </select>
      </div>
      <button
        class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
      >
        Gø!
      </button>
    </div>
  </form>
</div>
`;
